src/Cli/dotnet/commands/dotnet-workload/update/WorkloadUpdateCommand.cs

```diff
@@ -46,71 +49,128 @@ internal class WorkloadUpdateCommand : InstallingWorkloadCommand
+                                recorder.HistoryRecord.WorkloadArguments = new List<string>();
+                                foreach (var (id, version, featureBand) in rollbackFileContents)
+                                {
+                                    var idString = id.ToString();
+                                    recorder.HistoryRecord.WorkloadArguments.Add(idString);
+                                    rollbackContents[idString] = $"{version}/{featureBand}";
+                                }
+
+                                recorder.HistoryRecord.RollbackFileContents = rollbackContents;
+                            }
+
+                            UpdateWorkloads(recorder, _includePreviews, string.IsNullOrWhiteSpace(_fromCacheOption) ? null : new DirectoryPath(_fromCacheOption), rollbackFileContents);
+                        }
+                        catch (Exception e)
+                        {
+                            // Don't show entire stack trace
+                            throw new GracefulException(string.Format(LocalizableStrings.WorkloadUpdateFailed, e.Message), e, isUserError: false);
+                        }
+                    });
                 }
             }
-
-            _workloadInstaller.Shutdown();
+            finally
+            {
+                _workloadInstaller.Shutdown();
+            }
+            
             return _workloadInstaller.ExitCode;
         }
 
-        public void UpdateWorkloads(bool includePreviews = false, DirectoryPath? offlineCache = null)
+        public void UpdateWorkloads(WorkloadHistoryRecorder recorder = null, bool includePreviews = false, DirectoryPath? offlineCache = null, IEnumerable<(ManifestId id, ManifestVersion version, SdkFeatureBand featureBand)> rollbackFileContents = null)
         {
             Reporter.WriteLine();
 
             var workloadIds = GetUpdatableWorkloads();
+            if (recorder is not null)
+            {
+                recorder.HistoryRecord.WorkloadArguments = workloadIds.Select(id => id.ToString()).ToList();
+            }
+
             _workloadManifestUpdater.UpdateAdvertisingManifestsAsync(includePreviews, offlineCache).Wait();
 
-            var useRollback = !string.IsNullOrWhiteSpace(_fromRollbackDefinition);
+            var useRollbackOrHistory = !string.IsNullOrWhiteSpace(_fromRollbackDefinition) || !string.IsNullOrWhiteSpace(_fromHistorySpecified);
 
-            var manifestsToUpdate = useRollback ?
-                _workloadManifestUpdater.CalculateManifestRollbacks(_fromRollbackDefinition) :
-                _workloadManifestUpdater.CalculateManifestUpdates().Select(m => m.ManifestUpdate);
+            var manifestsToUpdate = CalculateManifestUpdates(recorder);
 
-            UpdateWorkloadsWithInstallRecord(_sdkFeatureBand, manifestsToUpdate, useRollback, offlineCache);
+            UpdateWorkloadsWithInstallRecord(_sdkFeatureBand, manifestsToUpdate, useRollbackOrHistory, offlineCache);
```

Comment:
Nit: This should make it a bit clearer how the parameter is being used.

```suggestion
            UpdateWorkloadsWithInstallRecord(_sdkFeatureBand, manifestsToUpdate, shouldUpdateInstallState: useRollbackOrHistory, offlineCache);
```